File: Practica3/QytetetRuby/pkg/QytetetRuby-0.0.1/lib/controlador_qytetet.rb
# To change this license header, choose License Headers in Project Properties.
# To change this template file, choose Tools | Templates
# and open the template in the editor.
require_relative 'qytetet.rb'
require_relative 'casilla.rb'
require_relative 'dado.rb'
require_relative 'jugador.rb'
require_relative 'metodo_salir_carcel.rb'
require_relative 'sorpresa.rb'
require_relative 'tablero.rb'
require_relative 'tipo_casilla.rb'
require_relative 'tipo_sorpresa.rb'
require_relative 'titulo_propiedad.rb'

#require_relative 'ModeloQytetet'

require_relative 'vista_textual_qytetet'

include ModeloQytetet

module InterfazTextualQytetet
  class ControladorQytetet  

    def initialize
      @vista = VistaTextualQytetet.new
      @jugador = nil
      @casilla = nil
      @juego = Qytetet.instance
    end
    
    def desarrollo_juego()
      tipo_casilla=0
      bancarrota = false
      
      while(!bancarrota)
          puts "El jugador actual es #{@jugador.nombre} con saldo #{@jugador.saldo} y se encuentra en la casilla #{@casilla}"
          
        if @jugador.encarcelado
            puts "estas en la carcel tienes dos metodos : tirar el dado para salir (escribe dado) o pagar libertad ( escribe libertad )"
          # metodo de lectura
          opcion = @vista.menu_salir_carcel
          salir=false
          if opcion == 0            
              salir = @juego.intentar_salir_carcel(MetodoSalirCarcel::TIRANDODADO)
          elsif opcion == 1
              salir = @juego.intentar_salir_carcel(MetodoSalirCarcel::PAGANDOLIBERTAD)
          end
          if !salir
              puts "ohhhh no has salido , pierdes turno"
              @jugador = @juego.siguiente_jugador
              #@jugador = @juego.jugador_actual
          else
              puts "felicidades sales de la carcer ya puedes seguir robando"
          end
        end
        if !@jugador.encarcelado
            salir = false
            puts "tirandooo el dadooooo"
            no_tiene_propietario = @juego.jugar()
            @casilla = @jugador.casilla_actual
            puts "El jugador actual es #{@jugador.nombre} con saldo #{@jugador.saldo} y se encuentra en la casilla #{@casilla}"
            if @casilla.tipo == TipoCasilla::SORPRESA                            
              aplicar_sorpresa = @juego.aplicar_sorpresa
              puts @juego.carta_actual
              
              if @jugador.saldo > 0
                bancarrota = false
              else 
                bancarrota = true
              end
            elsif @casilla.tipo == TipoCasilla::CARCEL
                puts "Estas en la carcel bien por el juez o sorpresa pierdes turno"
                @juego.siguiente_jugador
                salir=true
            else
              if !no_tiene_propietario
                recibido = @vista.elegir_quiero_comprar
                
                if (recibido == 0)
                  comprada = @juego.comprar_titulo_propiedad
                  if comprada
                    puts "Has conseguido comprarla!! Tu saldo es #{@jugador.saldo}"
                  else
                    puts "Tu saldo es insuficiente, no consigues comprarla"
                  end 
                  
                elsif (recibido == 1)
                  puts "Has decidido no comprar la propiedad"
                end
                
              else
                puts "Ohhh te toca pagar el alquiler y te queda de saldo #{@jugador.saldo}"
              end
            
            end
            
            while(!salir)
              recibido = @vista.menu_gestion_inmobiliaria
              case recibido
              when 0
                @juego.siguiente_jugador
                #@jugador = @juego.jugador_actual
                salir=true
              when 1
                if @jugador.tengo_propiedades
                  mias = Array.new
                  mias = @juego.casillas_propiedad_jugador
                  puts "estas son tus casillas para edificar : "
                  for i in 0..mias.size()-1
                    puts "en la posicion #{i} #{mias.at(i)}"
                  end
                  puts "\n Elige un numero de opcion: "
                  captura = gets.chomp
                  
                  puedo_construir = @juego.edificar_casa(mias.at(Integer(captura)))
                  if puedo_construir
                    puts "felicidades has podido construir una casa"
                    puts mias.at(Integer(captura))
                  else
                    puts "ohhhh no has podido construir una casa"
                  end
                end
                when 2
                  if @jugador.tengo_propiedades
                    mias = Array.new
                    mias = @juego.casillas_propiedad_jugador
                    puts "estas son tus casillas para edificar : "
                    for i in 0..mias.size()-1
                      puts "en la posicion #{i} #{mias.at(i)}"
                    end
                    puts "\n Elige un numero de opcion: "
                    captura = gets.chomp

                    puedo_construir = @juego.edificar_hotel(mias.at(Integer(captura)))
                    if puedo_construir
                      puts "felicidades has podido construir un hotel"
                      puts mias.at(Integer(captura))
                    else
                      puts "ohhhh no has podido construir un hotel, bien por falta de saldo o por no tener previamente 4 casas construidas"
                    end
                  end
                when 3
                  if @jugador.tengo_propiedades()
                    mias = Array.new
                    mias = @juego.casillas_propiedad_jugador()
                    puts "estas son tus casillas para vender: "
                    for i in 0..mias.size()-1
                      puts "en la posicion #{i} #{mias.at(i)}"
                    end
                    puts "\n Elige un numero de opcion: "
                    captura = gets.chomp

                    vendida = @juego.vender_propiedad(mias.at(Integer(captura)))
                    if vendida
                      puts "felicidades la has vendido"
                      puts mias.at(Integer(captura))
                    else
                      puts "oh no has podido vender"
                    end
                  else
                    puts "No tienes propiedades para vender"
                  end
                  @juego.siguiente_jugador()
                  
              when 4
                if @jugador.tengo_propiedades()
                  mias = Array.new
                  mias = @juego.casillas_propiedad_jugador()

                  puts "estas son tus casillas para hipotecar dale al numero para el que quieras hipotecar: "
                  for i in 0..mias.size()-1
                    puts "en la posicion #{i} #{mias.at(i)}"
                  end
                  puts "\n Elige un numero de opcion: "
                  captura = gets.chomp

                  hipotecada = @juego.hipotecar_propiedad(mias.at(Integer(captura)))
                  if hipotecada
                    puts "has conseguido hipotecarla"
                    puts mias.at(Integer(captura))
                  else
                    puts "no has conseguido hipotecarla"
                  end
                else
                  puts "No tienes propiedades para hipotecar"  
                end
                
               when 5
                if @jugador.tengo_propiedades()
                  mias = Array.new
                  mias = @juego.casillas_propiedad_jugador()                            
                  puts "estas son tus casillas para cancelar dale al numero para el que quieras cancelar: "
                  for i in 0..mias.size()-1
                    puts "en la posicion #{i} #{mias.at(i)}"
                  end
                  puts "\n Elige un numero de opcion: "
                  captura = gets.chomp

                  cancelada = @juego.cancelar_hipoteca(mias.at(Integer(captura)))
                  if cancelada
                    puts "has conseguido cancelar la hipoteca"
                    puts mias.at(Integer(captura))
                  else
                    puts "no has conseguido cancelar la hipoteca"
                  end
                else
                  puts "No tienes propiedades para cancelar" 
                end                        
              end
            end
          end
          @jugador = @juego.jugador_actual
        end
      
      #Fin del while
      if(bancarrota)
        puts "Se ha acabado el juego"
      end
    end
    
    def inicializacion_juego()
      nombres = Array.new
      nombres = @vista.obtener_nombre_jugadores
      @juego.inicializar_juego(nombres)
      @jugador = @juego.jugador_actual
      @casilla = @jugador.casilla_actual
      desarrollo_juego()
      
    end
       
    def elegir_propiedad(propiedades)
      
    end
    
    
    
    private :elegir_propiedad
    
    public
    def self.main
      prueba = ControladorQytetet.new
      prueba.inicializacion_juego
     
    end
    
  end
  ControladorQytetet.main
end

File: Practica3/QytetetJava/src/InterfazTextualQytetet/ControladorQytetet.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package InterfazTextualQytetet;

import java.util.ArrayList;
import java.util.Scanner;
import modeloqytetet.*;

/**
 *
 * @author jopoku
 */
public class ControladorQytetet {
    
    
    Jugador jugador;
    Casilla casilla;
    Qytetet juego = Qytetet.getInstance();
    VistaTextualQytetet vista = new VistaTextualQytetet();
    
    public void inicializarJuego(){
        
        ArrayList<String> nombres = vista.obtenerNumeroJugadores();
        juego.inicializarJuego(nombres);
        jugador = juego.getJugadorActual();
        casilla = jugador.getCasillaActual();
        desarrolloJuego();
        
    }
    
    private Casilla elegirCasilla(ArrayList<Casilla> propiedades){return null;
    }
    
    public void desarrolloJuego(){
        int tipo_Casilla = 0;
        boolean bancarrota=true;
       if(jugador.getSaldo() > 0){
            bancarrota= false;
       }
        while(!bancarrota){
        System.out.println("El jugador actual es : " + jugador.getNombre() + " con el saldo : "+ jugador.getSaldo() +" y la casilla en la que esta es : "+ jugador.getCasillaActual());
        
        if(jugador.getEncarcelado()){
            System.out.println("Estas en la carcel , tienes dos metodos : tirar el dado(escribe dado) o pagar la libertad(escribe libertad)");
            Scanner sc = new Scanner(System.in);
            String lectura = sc.nextLine();
            boolean salir=false;
            if(lectura.endsWith("dado")){
               salir = juego.intentarSalirCarcel(MetodoSalirCarcel.TIRANDODADO);
            }else if(lectura.endsWith("libertad")){
                salir = juego.intentarSalirCarcel(MetodoSalirCarcel.PAGANDOLIBERTAD);
            }
            if(!salir){
                System.out.println("OHHHHHHHHHH no has salido , pierdes el turno");
                jugador = juego.siguienteJugador();
            }else {
                System.out.println("Felicidades has salido de la carcel!!!!!! ya puedes seguir robando");
            }
        }
        if(!jugador.getEncarcelado()){ 
                //System.out.println("El jugador actual es : " + jugador.getNombre() + " y la casilla en la que esta es : "+ jugador.getCasillaActual());
                System.out.println("tirandooooo dadoooo .......");
                boolean notienepropietario = juego.jugar();
                casilla = jugador.getCasillaActual();
                System.out.println("El jugador actual es : " + jugador.getNombre());
                if(casilla.getTipo() == TipoCasilla.SORPRESA){
                    tipo_Casilla=1;
                     System.out.println(juego.getCartaActual());
                    boolean aplicarSorpresa = juego.aplicarSorpresa();
                     //System.out.println("El jugador actual es : " + jugador.getNombre() + " y la casilla en la que esta es : "+ casilla);
                     if(jugador.getSaldo() > 0){
                        bancarrota= false;
                        }else{
                         bancarrota = true;
                     }
                }
                else if(casilla.getTipo() == TipoCasilla.CARCEL){
                    System.out.println("ohhhhh has caido en la casilla del juez o te ha tocado una carta sorpresa y te vas a la carceel");
                    tipo_Casilla=1;
                }
                else{
                    tipo_Casilla=2;
                    System.out.println(casilla);
                    if(!notienepropietario){
                        int recibido = vista.menuGestionInmobiliaria(3);
                         switch(recibido){
                    case 0: boolean comprada = juego.comprarTituloPropiedad();
                        if(comprada){
                            System.out.println("Has conseguido comprarla" + " y te queda de saldo : " + jugador.getSaldo());
                        }else{
                            System.out.println("ohhh no has podido comprarla");
                            tipo_Casilla=1;
                        }
                            break;
                    case 1: tipo_Casilla=1; 
                        break;
                         
                    }
                    }else {
                          System.out.println("Ohhhh te toca pagar alquiler y te quedas con : " + jugador.getSaldo());
                                                   
                    }
                }
                boolean salir=false;
                while(!salir){
                    int recibido = vista.menuGestionInmobiliaria(tipo_Casilla);
                    switch(recibido ){
                    case 0: juego.siguienteJugador();
                            salir=true;
                            break;
                    case 1: if(jugador.tengoPropiedades()){
                            ArrayList<Casilla> mias = juego.casillasPropiedadJugador();
                            System.out.println("Estas son tus casillas para edificar, dale al numero para el que quieras edificar");
                            for (int i = 0; i < mias.size(); i++) {
                                System.out.println( "en la posicion " + i + " " +mias.get(i).getTitulo().getNombre());
                             }
                            Scanner sc = new Scanner(System.in);
                            String lectura = sc.nextLine(); 
                            boolean puedoConstruir = juego.edificarCasa(mias.get(Integer.parseInt(lectura)));
                                if(puedoConstruir){
                                    System.out.println("Felicidades has construido una casa");
                                    System.out.println(mias.get(Integer.parseInt(lectura)));
                                }else{
                                    System.out.println("Oh no has podido construir la casa");
                                }
                            }
                            else{
                                System.out.println("No tienes propiedades donde edificar");
                            }
     
                            break;
                    case 2: if(jugador.tengoPropiedades()){
                            ArrayList<Casilla> mias = juego.casillasPropiedadJugador();
                            System.out.println("Estas son tus casillas para edificar, dale al numero para el que quieras edificar");
                            for (int i = 0; i < mias.size(); i++) {
                                System.out.println( "en la posicion " + i + " " +mias.get(i).getTitulo().getNombre());
                             }
                            Scanner sc = new Scanner(System.in);
                            String lectura = sc.nextLine(); 
                            boolean puedoConstruir = juego.edificarHotel(mias.get(Integer.parseInt(lectura)));
                                if(puedoConstruir){
                                    System.out.println("Felicidades has construido un hotel");
                                    System.out.println(mias.get(Integer.parseInt(lectura)));
                                }else{
                                    System.out.println("Oh no has podido construir el hotel, bien por falta de saldo o porque no tienes 4 casas construidas antes");
                                }
                            }
                            else{
                                System.out.println("No tienes propiedades donde edificar");
                            }
                            break;
                    case 3: if(jugador.tengoPropiedades()){
                            ArrayList<Casilla> mias = juego.casillasPropiedadJugador();
                            System.out.println("Estas son tus casillas para vender, dale al numero para el que quieras vender");
                            for (int i = 0; i < mias.size(); i++) {
                                System.out.println( "en la posicion " + i + " " +mias.get(i).getTitulo().getNombre());
                             }
                            Scanner sc = new Scanner(System.in);
                            String lectura = sc.nextLine(); 
                               boolean vendida = juego.venderPropiedad(mias.get(Integer.parseInt(lectura)));
                               if(vendida){
                                    System.out.println("Felicidades la has vendido");
                                }else{
                                    System.out.println("Oh no has podido vender");
                                }
                            }
                            else{
                                System.out.println("No tienes propiedades para vender");
                            }
                            juego.siguienteJugador();
                            break;
                    case 4: if(jugador.tengoPropiedades()){
                            ArrayList<Casilla> mias = juego.casillasPropiedadJugador();
                            System.out.println("Estas son tus casillas para hipotecar, dale al numero correspondiente al que quieras hipotecar");
                            for (int i = 0; i < mias.size(); i++) {
                                System.out.println( "en la posicion " + i + " " +mias.get(i).getTitulo().getNombre());
                             }
                            Scanner sc = new Scanner(System.in);
                            String lectura = sc.nextLine(); 
                             boolean hipotecada =juego.hipotecarPropiedad(mias.get(Integer.parseInt(lectura)));
                                if(hipotecada){
                                    System.out.println("has conseguido hipotecarla");
                                }else {
                                    System.out.println("no has conseguido hipotecarla");
                                }
                                
                            }
                            else{
                                System.out.println("No tienes propiedades para hipotecar");
                            }
                            break;
                    case 5: if(jugador.tengoPropiedades()){
                            ArrayList<Casilla> mias = juego.casillasPropiedadJugador();
                            System.out.println("Estas son tus casillas para cancelar, dale al numero correspondiente al que quieras cancelar");
                            for (int i = 0; i < mias.size(); i++) {
                                System.out.println( "en la posicion " + i + " " + mias.get(i).getTitulo().getNombre());
                             }
                            Scanner sc = new Scanner(System.in);
                            String lectura = sc.nextLine(); 
                             boolean hipotecada =juego.cancelarHipoteca(mias.get(Integer.parseInt(lectura)));
                                if(hipotecada){
                                    System.out.println("has conseguido cancerla la hipoteca");
                                }else {
                                    System.out.println("no has conseguido cancelar la hipoteca");
                                }
                                
                            }
                            else{
                                System.out.println("No tienes propiedades para cancelar hipotecar");
                            }
                            break;
                         
                     }
                }
               
                
             }  
              jugador = juego.getJugadorActual();
            }
       
        
        //fin del while
        if(bancarrota){
            System.out.println("Se ha acabado el juego.");
        }
    }
    
    public static void main(String[] args) {
        ControladorQytetet prueba = new ControladorQytetet();
        prueba.inicializarJuego();
    }
}
